From 656b8f7af3cc106c79cf76755c982ccfdca6a9fb Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Wed, 20 Oct 2021 22:02:14 +0200
Subject: [PATCH 121/122] clk: meson: mpll: consider sdm_en in .recalc_rate and
 .set_rate

The SDM value is the fractial part of an MPLL clock on Amlogic Meson
SoCs. This value is only used by the hardware when the sdm_en bit is set
to 1.

Make .recalc_rate read this bit and (depending on it's value) ignore or
honor the SDM value during rate calculation. Also update .set_rate to
toggle the bit depending on whether the SDM value is zero or non-zero.

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
---
 drivers/clk/meson/clk-mpll.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/clk/meson/clk-mpll.c b/drivers/clk/meson/clk-mpll.c
index 20255e129b37..b843e101d1ba 100644
--- a/drivers/clk/meson/clk-mpll.c
+++ b/drivers/clk/meson/clk-mpll.c
@@ -77,10 +77,13 @@ static unsigned long mpll_recalc_rate(struct clk_hw *hw,
 {
 	struct clk_regmap *clk = to_clk_regmap(hw);
 	struct meson_clk_mpll_data *mpll = meson_clk_mpll_data(clk);
-	unsigned int sdm, n2;
+	unsigned int n2, sdm_en, sdm = 0;
 	long rate;
 
-	sdm = meson_parm_read(clk->map, &mpll->sdm);
+	sdm_en = meson_parm_read(clk->map, &mpll->sdm_en);
+	if (sdm_en)
+		sdm = meson_parm_read(clk->map, &mpll->sdm);
+
 	n2 = meson_parm_read(clk->map, &mpll->n2);
 
 	rate = rate_from_params(parent_rate, sdm, n2);
@@ -124,6 +127,9 @@ static int mpll_set_rate(struct clk_hw *hw,
 	/* Set the fractional part */
 	meson_parm_write(clk->map, &mpll->sdm, sdm);
 
+	/* Enable or disable the fractional part as needed */
+	meson_parm_write(clk->map, &mpll->sdm_en, sdm > 0 ? 1 : 0);
+
 	/* Set the integer divider part */
 	meson_parm_write(clk->map, &mpll->n2, n2);
 
@@ -144,9 +150,6 @@ static int mpll_init(struct clk_hw *hw)
 		regmap_multi_reg_write(clk->map, mpll->init_regs,
 				       mpll->init_count);
 
-	/* Enable the fractional part */
-	meson_parm_write(clk->map, &mpll->sdm_en, 1);
-
 	/* Set spread spectrum if possible */
 	if (MESON_PARM_APPLICABLE(&mpll->ssen)) {
 		unsigned int ss =
-- 
2.25.1

